iroh-bytes/src/store/bao_file.rs

```diff
@@ -0,0 +1,1238 @@
+                    if let Some(offset) = tree.pre_order_offset(parent.node) {
+                        let o0 = offset * 64;
+                        self.outboard
+                            .write_all_at(o0, parent.pair.0.as_bytes().as_slice())?;
+                        self.outboard
+                            .write_all_at(o0 + 32, parent.pair.1.as_bytes().as_slice())?;
+                    }
+                }
+                BaoContentItem::Leaf(leaf) => {
+                    let o0 = leaf.offset.0;
+                    // divide by chunk size, multiply by 8
+                    let index = (leaf.offset.0 >> (tree.block_size().0 + 10)) << 3;
+                    tracing::debug!(
+                        "write_batch f={:?} o={} l={}",
+                        self.data,
+                        o0,
+                        leaf.data.len()
+                    );
+                    self.data.write_all_at(o0, leaf.data.as_ref())?;
+                    let size = tree.size().0;
+                    self.sizes.write_all_at(index, &size.to_le_bytes())?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn read_data_at(&self, offset: u64, len: usize) -> io::Result<Bytes> {
+        read_to_end(&self.data, offset, len)
+    }
+
+    fn read_outboard_at(&self, offset: u64, len: usize) -> io::Result<Bytes> {
+        read_to_end(&self.outboard, offset, len)
+    }
+}
+
+/// The storage for a bao file. This can be either in memory or on disk.
+#[derive(Debug)]
+pub(crate) enum BaoFileStorage {
+    /// The entry is incomplete and in memory.
+    ///
+    /// Since it is incomplete, it must be writeable.
+    ///
+    /// This is used mostly for tiny entries, <= 16 KiB. But in principle it
+    /// can be used for larger sizes.
+    ///
+    /// Incomplete mem entries are *not* persisted at all. So if the store
+    /// crashes they will be gone.
+    IncompleteMem(MutableMemStorage),
+    /// The entry is incomplete and on disk.
+    IncompleteFile(FileStorage),
+    /// The entry is complete. Outboard and data can come from different sources
+    /// (memory or file).
+    ///
+    /// Writing to this is a no-op, since it is already complete.
+    Complete(CompleteMemOrFileStorage),
+}
+
+impl Default for BaoFileStorage {
+    fn default() -> Self {
+        BaoFileStorage::Complete(Default::default())
+    }
+}
+
+impl BaoFileStorage {
+    /// Take the storage out, leaving an empty storage in its place.
+    ///
+    /// Be careful to put somethign back in its place, or you will lose data.
+    #[cfg(feature = "file-db")]
+    pub fn take(&mut self) -> Self {
+        std::mem::take(self)
+    }
+
+    /// Create a new mutable mem storage.
+    pub fn incomplete_mem() -> Self {
+        Self::IncompleteMem(Default::default())
+    }
+
+    /// Call sync_all on all the files.
+    fn sync_all(&self) -> io::Result<()> {
+        match self {
+            Self::Complete(_) => Ok(()),
```

Comment:
why is there no need to propagate the `sync_all` to the potential files in this?

Reply:
A complete file is immutable. Both the data and the outboard. So any attempts to write to a complete file are noops that are just there to make things simpler.

Reply:
maybe change the doc comment then to sth like `propagate to all mutable files`